### PR TITLE
Add tune threshold functionality

### PR DESCRIPTION
**Pull request by nsorros**
This PR implements tuning threshold for our tagger inspired from the paper https://doi.org/10.1016/j.patcog.2013.01.012 but modified to be efficient for MeSH. Modifications are:

- Tune threshold on a sample of the test set
- Try a fixed number of thresholds in the continuous range 0 to 1. For example if we want to try 5 thresholds they would be 0 0.2 0.4 0.6 0.8
- Init threshold value instead of starting from an unnecessarily bad init value which requires more iterations

### CODE REVIEW

**Comment by nsorros** [comment]
[Tune threshold for each mesh term or label term independently to maximise f1](https://trello.com/c/wWyX0lzP/1136-tune-threshold-for-each-mesh-term-or-label-term-independently-to-maximise-f1)

File: grants_tagger/tune_threshold.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    return optimal_thresholds
+
+
+def tune_threshold(data_path, model_path, label_binarizer_path, thresholds_path, sample_size=None, nb_thresholds=None, init_threshold=None):
+    with open(label_binarizer_path, "rb") as f:
+        label_binarizer = pickle.loads(f.read())
+
+    X, Y, _ = load_data(data_path, label_binarizer)
+    _, X_test, _, Y_test = train_test_split(X, Y, random_state=42)
```
Comment (lizgzil) [inline grants_tagger/tune_threshold.py:98]: is this the same split you used when training the model? if it's not, I guess it might overfit your thresholds to the training data?

File: grants_tagger/evaluate_model.py [hunk 2]
```diff
@@ -17,20 +17,26 @@
 from grants_tagger.predict import predict_proba_ensemble_tfidf_svm_bert, load_model
 
 
-def predict(X_test, model_path, nb_labels, threshold):
+def predict(X_test, model_path, nb_labels=None, threshold=None, return_probabilities=False):
```
Comment (lizgzil) [inline grants_tagger/evaluate_model.py:20]: perhaps `threshold` should be `thresholds`?

File: grants_tagger/tune_threshold.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    optimal_threshold_i = current_threshold_i
+
+    Y_pred = Y_pred_proba > candidate_thresholds
+    Y_pred = csr_matrix(Y_pred)
+    cm = multilabel_confusion_matrix(Y_test, Y_pred)
+
+    if nb_thresholds:
+        candidate_thresholds_i = [t/nb_thresholds for t in range(1, nb_thresholds)]
+    else:
+        candidate_thresholds_i = Y_pred_proba[:, label_i]
```
Comment (lizgzil) [inline grants_tagger/tune_threshold.py:38]: Might be worth `np.unique(Y_pred_proba[:, label_i])`?

File: grants_tagger/tune_threshold.py (hunk 1 above)
Comment (nsorros) [reply grants_tagger/tune_threshold.py:98]: this is a good point, i think i might have another function that splits but essentially that's what it does under the hood. i will check and replace

File: grants_tagger/tune_threshold.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger/tune_threshold.py:38]: good idea

File: grants_tagger/evaluate_model.py (hunk 2 above)
Comment (nsorros) [reply grants_tagger/evaluate_model.py:20]: tricky one as you can pass in both a value and a list, not sure what the right name in these cases

**Comment by lizgzil** [comment]
also `make test` passed for me 👍

**Comment by nsorros** [comment]
> also `make test` passed for me 👍

that's some good news

File: grants_tagger/tune_threshold.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger/tune_threshold.py]: 8b1752b

File: grants_tagger/tune_threshold.py (hunk 1 above)
Comment (nsorros) [reply grants_tagger/tune_threshold.py]: 59aa03e

File: grants_tagger/evaluate_model.py [hunk 4]
```diff
@@ -17,20 +17,26 @@
...
+        if return_probabilities:
+            Y_pred = Y_pred_proba
+        else:
+            Y_pred = Y_pred_proba > threshold
     elif "disease_mesh_cnn" in model_path:
-        Y_pred = predict_cnn(X_test, model_path, threshold)
+        Y_pred = predict_cnn(X_test, model_path, threshold, return_probabilities)
     elif "disease_mesh_tfidf" in model_path:
-        Y_pred = predict_tfidf_svm(X_test, model_path, nb_labels, threshold)
+        Y_pred = predict_tfidf_svm(X_test, model_path, nb_labels, threshold, return_probabilities)
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:32]: actually this is a bug well spotted. added a test as well. fixed in 3516143

File: grants_tagger/evaluate_model.py [hunk 5]
```diff
@@ -17,20 +17,26 @@
...
+def predict(X_test, model_path, nb_labels=None, threshold=None, return_probabilities=False):
     # comma indicates ensemble of more than one models
     if "," in model_path:
         model_paths = model_path.split(",")
         Y_pred_proba = predict_proba_ensemble_tfidf_svm_bert(X_test, model_paths)
-        Y_pred = Y_pred_proba > threshold
+        if return_probabilities:
+            Y_pred = Y_pred_proba
+        else:
+            Y_pred = Y_pred_proba > threshold
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:28]: good point, will default to None

[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:28]: 2a49f37

**Review by lizgzil — APPROVED** [review]
